Unobtrusive feature

share-with-me opened the pull request.

share-with-me commented:
@sushain97 , `newline-per-chained-call` means something like the following?? What about indentation? Can I put a new line at each function call?

```
$('#installationNotice').removeClass('hide').fadeIn('slow').
                                                     delay(messageDuration).
                                                     fadeOut('slow', hideUnobtrusiveWarning);
```

share-with-me commented:
Um, @sushain97 , Could you review my above comment on this? So i could get Travis pass the code!

sushain97 commented:
> @sushain97 , newline-per-chained-call means something like the following?? What about indentation? Can I put a new line at each function call?

Each chained call goes on a separate line with the dots at the start of the line.

All the rules are documented on ESLint, e.g. http://eslint.org/docs/rules/newline-per-chained-call.

share-with-me commented:
@sushain97 Thanks for the link! Um, I need to fix the warnings too? :P . Errors should do right? :P

sushain97 commented:
> @sushain97 Thanks for the link! Um, I need to fix the warnings too? :P . Errors should do right? :P

Nope. I would make all the warnings errors as well but that is tricky to do with `sass-lint` as far as I can tell. No warnings or errors ==> mergeable.

sushain97 commented on assets/strings/eng.json:
```diff
@@ -61,5 +61,7 @@
     "Downloads_Para": "Current versions of the <b>Apertium toolbox</b> as well as of <b>language-pair data</b> are available from <a href='https://sourceforge.net/projects/apertium/' target='_blank'>the SourceForge page</a>. Installation instructions for Apertium on all major platforms are provided on the <a href='http://wiki.apertium.org/wiki/Installation' target='_blank'>Wiki's Installation pag ...
-    "Contact_Para": "<h4>IRC channel</h4> <p>The quickest way to contact us is by joining our <a href='http://wiki.apertium.org/wiki/IRC' target='_blank'><abbr title='Internet Relay Chat'>IRC</abbr> channel, #apertium</a> at irc.freenode.net, where users and developers of Apertium meet. You don't need an IRC client; you can use <a href='http://webchat.freenode.net/?channels=apertium'>freenode web ...
+    "Contact_Para": "<h4>IRC channel</h4> <p>The quickest way to contact us is by joining our <a href='http://wiki.apertium.org/wiki/IRC' target='_blank'><abbr title='Internet Relay Chat'>IRC</abbr> channel, #apertium</a> at irc.freenode.net, where users and developers of Apertium meet. You don't need an IRC client; you can use <a href='http://webchat.freenode.net/?channels=apertium'>freenode web ...
+    "Installation_Notice_Title": "Inconvenience",
```
`Inconvenience` => `Install Apertium`

sushain97 commented on assets/strings/eng.json:
```diff
@@ -61,5 +61,7 @@
-    "Contact_Para": "<h4>IRC channel</h4> <p>The quickest way to contact us is by joining our <a href='http://wiki.apertium.org/wiki/IRC' target='_blank'><abbr title='Internet Relay Chat'>IRC</abbr> channel, #apertium</a> at irc.freenode.net, where users and developers of Apertium meet. You don't need an IRC client; you can use <a href='http://webchat.freenode.net/?channels=apertium'>freenode web ...
+    "Contact_Para": "<h4>IRC channel</h4> <p>The quickest way to contact us is by joining our <a href='http://wiki.apertium.org/wiki/IRC' target='_blank'><abbr title='Internet Relay Chat'>IRC</abbr> channel, #apertium</a> at irc.freenode.net, where users and developers of Apertium meet. You don't need an IRC client; you can use <a href='http://webchat.freenode.net/?channels=apertium'>freenode web ...
+    "Installation_Notice_Title": "Inconvenience",
+    "Installation_Notice": "We are highly sorry for the inconvenience caused but our server appears loaded now. In order to set apertium locally, please refer to this <a href='http://wiki.apertium.org/wiki/Apertium-apy' target='_blank'> link</a>."
```
`"Are you experiencing slow responses? Our servers might be overloaded. Learn how to <a href="http://wiki.apertium.org/wiki/Installation">install Apertium locally</a>."`

sushain97 commented on assets/js/util.js:
```diff
@@ -293,13 +300,82 @@ function callApy(options, endpoint) {
+    var storedLastNRequestsDuration = [];
+    var demoThreshold = 10000; // for demo purposes only
+    try {
+        window.sessionStorage;
```
I thought about this a bit more and I don't think these need to be stored in `sessionStorage`. 

From our conversation on the PR:

> My current understanding is, for the current session ie until user closes the tab/window, I store all the page load times in a session variable and then calculate the average?

This is fine. Simple and easy.

From this I assumed that session storage persisted only through the life of the tab but it seems to be the life of the browser instead. I think it makes more sense (and drastically simplifies the code) for this to just be per page. We should also make sure to only display the message to the user once. We don't want them to keep seeing it.

share-with-me replied:
> From this I assumed that session storage persisted only through the life of the tab but it seems to be the life of the browser instead.

Yes, a bit more reading makes me realize this!

>  I think it makes more sense (and drastically simplifies the code) for this to just be per page.

Oh! Um, that means I could just store these values in an `array`. And everytime the user refreshes the page, all of this starts afresh? 

> We should also make sure to only display the message to the user once. We don't want them to keep seeing it.

Yes, we do not want to annoy them!!

sushain97 commented on assets/js/util.js:
```diff
@@ -11,6 +11,11 @@ var TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH = 768,
 
+var apyRequestStartTime,
+    INSTALLATION_NOTIFICATION_REQUESTS_BUFFER_LENGTH = 10,
+    INDIVIDUAL_DURATION_THRESHOLD = 2000,
```
Prefix these two variables with `INSTALLATION_NOTIFICATION_`

share-with-me commented:
> (travis and conflict)

On it

share-with-me commented:
@sushain97 , Um, `fa-times` is a `font-awesome` class. Travis is giving an error stating it should be made camelcase. Should I also add a custom class there and use that to style the element?

sushain97 commented:
> And then style using faTimes in CSS?

That's a good approach. An easier solution is to simply use the linter ignore syntax (look for it in other places in the file).

share-with-me commented:
>  An easier solution is to simply use the linter ignore syntax (look for it in other places in the file)

Thanks @sushain97 :D. Done!!!

sushain97 commented on assets/js/util.js:
```diff
@@ -11,6 +11,12 @@ var TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH = 768,
 
+var apyRequestStartTime, installationNoticeFlag = true, lastNRequestsDuration = [],
+    requestsMade = 0, cumulativeRequestsTime = 0,
+    INSTALLATION_NOTIFICATION_REQUESTS_BUFFER_LENGTH = 10,
```
Move the constants to the place where the rest of the constants are.

share-with-me replied:
Done

sushain97 commented on assets/js/util.js:
```diff
@@ -11,6 +11,12 @@ var TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH = 768,
     BACK_TO_TOP_BUTTON_ACTIVATION_HEIGHT = 300,
     APY_REQUEST_URL_THRESHOLD_LENGTH = 2000; // maintain 48 characters buffer for generated parameters
 
+var apyRequestStartTime, installationNoticeFlag = true, lastNRequestsDuration = [],
```
Use more descriptive names such as `installationNoticeShown`.

share-with-me replied:
Yes!

sushain97 commented on assets/js/util.js:
```diff
@@ -162,6 +171,7 @@ $(document).ready(function () {
         return false;
     });
 
+    $('#installationNotice').addClass('hide');
```
Why not just `.hide()`?

sushain97 commented:
Trying out this feature, I'd also like the notification to appear if a running request takes longer than the `INSTALLATION_NOTIFICATION_INDIVIDUAL_DURATION_THRESHOLD`. Also, we're switching around with our terminology. Should we use notice or notification...

sushain97 commented:
For the notification styling, you're on a good track. Try something a little closer to vanilla Bootstrap: http://getbootstrap.com/components/#alerts. Also, the link needs to be `target="_blank"`.

share-with-me commented:
> Also, we're switching around with our terminology. Should we use notice or notification...

According to me, `notification` makes more sense!

share-with-me commented:
> Trying out this feature, I'd also like the notification to appear if a running request takes longer than the INSTALLATION_NOTIFICATION_INDIVIDUAL_DURATION_THRESHOLD.

Makes sense! Modified the code :)

> Why not just .hide()?

I have added classes `hidden-xs hidden-sm` to HTML. This adds a property of `display: block !important` (for larger screen devices). Thus, simply using `hide()` or `show()` does not override these bootstrap inline styles. Adding and removing `hide` class overrides these limitations and displays and shows the `div`. Thus, I make use of `addClass('hide')` and `removeClass('hide')`. This [stackoverflow link](https://stackoverflow.com/questions/18568736/how-to-hide-element-using-twitter-bootstrap-3-and-show-it-using-jquery) discusses about how `hide, addClass and removeClass` can be used to achieve the desired functionality.

share-with-me commented:
>  Try something a little closer to vanilla Bootstrap: http://getbootstrap.com/components/#alerts.

Um, I have used `alert-warning`. Should I make use of `alert-info/success`?

sushain97 commented:
> Um, I have used alert-warning. Should I make use of alert-info/success?

Hm, either `warning` or `info`... Show me how both look?

share-with-me commented:
> Hm, either warning or info... Show me how both look?

Yes :D.

1. Warning:

![screenshot from 2017-06-25 00-26-21](https://user-images.githubusercontent.com/8894636/27511217-0b760aa4-593d-11e7-8ba1-69671042a67c.png)

2. Info:

![screenshot from 2017-06-25 00-27-51](https://user-images.githubusercontent.com/8894636/27511222-2491b222-593d-11e7-9fb7-c6340a9b161c.png)

3. Success:

![screenshot from 2017-06-25 00-28-31](https://user-images.githubusercontent.com/8894636/27511226-3bd89540-593d-11e7-9a9b-fcfda939d830.png)

sushain97 commented:
I'm honestly fine with either of the first two. Is there a reason you're using a large white border? Its seems a bit unsightly to me.

share-with-me commented:
> Is there a reason you're using a large white border?

Um, no! I just did not change the notification CSS :D. Have removed it now.

sushain97 commented on assets/js/util.js:
```diff
@@ -9,7 +9,14 @@ var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
-    APY_REQUEST_URL_THRESHOLD_LENGTH = 2000; // maintain 48 characters buffer for generated parameters
+    APY_REQUEST_URL_THRESHOLD_LENGTH = 2000, // maintain 48 characters buffer for generated parameters
+    INSTALLATION_NOTIFICATION_REQUESTS_BUFFER_LENGTH = 10,
+    INSTALLATION_NOTIFICATION_INDIVIDUAL_DURATION_THRESHOLD = 2000,
```
double these values

sushain97 commented on assets/js/util.js:
```diff
@@ -9,7 +9,14 @@ var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
+    INSTALLATION_NOTIFICATION_INDIVIDUAL_DURATION_THRESHOLD = 2000,
+    INSTALLATION_NOTIFICATION_CUMULATIVE_DURATION_THRESHOLD = 1500;
+
+var apyRequestStartTime, installationNotificationShown = false, isCallApyCompleted,
```
`isCallApyCompleted` -> `apyRequestTimeout`

sushain97 commented on assets/js/util.js:
```diff
@@ -45,6 +52,10 @@ function ajaxSend() {
 
 function ajaxComplete() {
     $('#loadingIndicator').hide();
+    if(installationNotificationShown === false) {
```
`!installationNotificationShown`

sushain97 commented on assets/js/util.js:
```diff
@@ -293,13 +305,63 @@ function callApy(options, endpoint) {
+}
+
+function displayUnobtrusiveWarning() {
+    var messageDuration = 10000;
```
Move this to a constant.

sushain97 commented on assets/js/util.js:
```diff
@@ -9,7 +9,14 @@ var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
+    INSTALLATION_NOTIFICATION_CUMULATIVE_DURATION_THRESHOLD = 3000;
+
+var apyRequestStartTime, installationNotificationShown = false, apyRequestTimeout,
+    lastNRequestsDuration = [], requestsMade = 0, cumulativeRequestsTime = 0, installationNotificationDuration = 10000;
```
This constant needs to be moved up as well.

sushain97 commented on assets/js/util.js:
```diff
@@ -293,13 +305,62 @@ function callApy(options, endpoint) {
     var requestUrl = window.location.protocol + window.location.hostname +
         window.location.pathname + '?' + $.param(requestOptions.data);
 
+    if(!installationNotificationShown) {
```
This is always going to be `true` when the page loads.

share-with-me replied:
Yes, because the notification is not yet shown! Are you suggesting that the value holds even when the user refreshes?

sushain97 replied:
Whoops, I mistook the context in which this gets run. Can we move this check into `displayUnobtrusiveWarning` (which we should rename to `displayInstallationNotification`) - you can make it an early return. Then it only needs to happen in one place.

sushain97 commented on assets/js/util.js:
```diff
@@ -293,13 +305,62 @@ function callApy(options, endpoint) {
+    if(!installationNotificationShown) {
+        apyRequestStartTime = Date.now();
+        apyRequestTimeout = setTimeout(function () {
+            displayUnobtrusiveWarning();
```
Isn't this going to run regardless of how long the request takes?

share-with-me replied:
Nope, I am clearing the `apyRequestTimeout` variable on `ajaxComplete`. Thus, if my AJAX request is completed before the INDIVIDUAL_THRESHOLD, the variable will be cleared and it will not execute.

sushain97 replied:
Wow, I completely missed that. I think it's time to go to bed!

share-with-me replied:
Hahahahahahhaha! yes, please sleep well :P

sushain97 commented on assets/js/util.js:
```diff
@@ -293,13 +306,62 @@ function callApy(options, endpoint) {
+    });
+}
+
+function hideUnobtrusiveWarning() {
```
(also change the name of this one)

share-with-me replied:
On it :D

share-with-me commented:
Done @sushain97  :D
